### PR TITLE
[Progression] call filter_stacktrace with a signed_url within utask_main

### PR DESCRIPTION
**Pull request by alhijazi**

### CODE REVIEW

File: src/clusterfuzz/_internal/google_cloud_utils/storage.py [hunk 1]
```diff
@@ -1170,3 +1170,30 @@ def get_signed_download_url(remote_path, minutes=SIGNED_URL_EXPIRATION_MINUTES):
...
+
+
+def get_blob_signed_upload_url():
+  """Returns a pair of (blob_name,signed_upload_url) to be used from utask_main
+  to upload blobs."""
+  bucket = blobs_bucket()
+  from clusterfuzz._internal.google_cloud_utils import blobs
+  blob_name = blobs.generate_new_blob_name()
+  gcs_path = get_cloud_storage_file_path(bucket, blob_name)
+  # keep generating paths until no collision is encountered.
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/google_cloud_utils/storage.py:1182]: Keep

File: src/clusterfuzz/_internal/google_cloud_utils/storage.py [hunk 2]
```diff
@@ -1170,3 +1170,30 @@ def get_signed_download_url(remote_path, minutes=SIGNED_URL_EXPIRATION_MINUTES):
   contents."""
   provider = _provider()
   return provider.sign_download_url(remote_path, minutes=minutes)
+
+
+def get_blob_signed_upload_url():
+  """Returns a pair of (blob_name,signed_upload_url) to be used from utask_main
+  to upload blobs."""
+  bucket = blobs_bucket()
+  from clusterfuzz._internal.google_cloud_utils import blobs
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/google_cloud_utils/storage.py:1179]: Don't do this please it introduces a circular import. Can you put this in blobs.py?

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 3]
```diff
@@ -111,6 +111,21 @@ def handle_progression_revision_list_error(
...
+def _cleanup_stacktrace_blob_from_storage(
+    uworker_output: uworker_msg_pb2.Output):
+  """Cleanup the blob created in preprocess if it wasn't used to store the
+  filterd stacktrace."""
+  if (uworker_output.HasField('progression_task_output') and
+      uworker_output.progression_task_output.last_tested_crash_stacktrace.
+      startswith(data_types.BLOBSTORE_STACK_PREFIX)):
+    return
+
+  assert uworker_output.uworker_input.progression_task_input.blob_name
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:123]: Hmmm...Do you think we should use an assert here instead of an exception.

Comment (alhijazi) [reply src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py]: Replaced with a `ValueError`.

File: src/clusterfuzz/_internal/google_cloud_utils/storage.py (hunk 2 above)
Comment (alhijazi) [reply src/clusterfuzz/_internal/google_cloud_utils/storage.py]: Done.

File: src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py [hunk 4]
```diff
@@ -111,6 +112,22 @@ def handle_progression_revision_list_error(
                                          'Failed to fetch revision list')
 
 
+def _cleanup_stacktrace_blob_from_storage(
+    uworker_output: uworker_msg_pb2.Output):
+  """Cleanup the blob created in preprocess if it wasn't used to store the
+  filterd stacktrace."""
+  if (uworker_output.HasField('progression_task_output') and
+      uworker_output.progression_task_output.last_tested_crash_stacktrace.
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py:120]: This is very hard to read. Make this a second condition afterwards or something please.

Comment (alhijazi) [reply src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py]: Done.

**Comment by alhijazi** [comment]
@jonathanmetzman Can we land this?

File: src/clusterfuzz/_internal/google_cloud_utils/blobs.py [hunk 5]
```diff
@@ -229,3 +229,20 @@ def get_signed_upload_url(blob_key):
...
+def get_blob_signed_upload_url():
+  """Returns a pair of (blob_name,signed_upload_url) to be used from utask_main
+  to upload blobs."""
+  bucket = storage.blobs_bucket()
+  blob_name = generate_new_blob_name()
+  gcs_path = storage.get_cloud_storage_file_path(bucket, blob_name)
+  # Keep generating paths until no collision is encountered.
+  while storage.get(gcs_path):
+    blob_name = generate_new_blob_name()
+    gcs_path = storage.get_cloud_storage_file_path(bucket, blob_name)
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/google_cloud_utils/blobs.py:243]: I think it's fine to just exception. The collisions are extremely unlikely to happen. We don't need to add code to deal with it.

**Review by jonathanmetzman — COMMENTED** [review]
LGTM with one nit

**Pull request merged**